src/sqlparser/src/parser.rs

```diff
@@ -899,6 +931,68 @@ impl Parser {
         }
     }
 
+    /// Parses an array expression `[ex1, ex2, ..]`
+    /// if `named` is `true`, came from an expression like  `ARRAY[ex1, ex2]`
+    pub fn parse_array_expr(&mut self, named: bool) -> Result<Expr, ParserError> {
+        self.increase_array_depth(1);
+        if let hash_map::Entry::Vacant(entry) = self.array_named_map.entry(self.peek_array_depth())
+        {
+            entry.insert(named);
+        } else if let Err(parse_err) = self.check_same_named_array(named) {
+            Err(parse_err)?
+        }
+
+        if self.peek_token() == Token::RBracket {
+            let _ = self.next_token(); // consume ]
+            self.decrease_array_depth(1);
+            Ok(Expr::Array(Array {
+                elem: vec![],
+                named,
+            }))
+        } else {
+            let exprs = self.parse_comma_separated(Parser::parse_expr)?;
+            self.expect_token(&Token::RBracket)?;
+            let ended_array_depth = self.peek_array_depth() + 1;
+            if self.array_named_map.contains_key(&ended_array_depth) {
+                self.array_named_map.remove(&ended_array_depth);
+            }
+            self.decrease_array_depth(1);
+            Ok(Expr::Array(Array { elem: exprs, named }))
+        }
+    }
+
+    fn check_same_named_array(&mut self, current_named: bool) -> Result<(), ParserError> {
+        match current_named {
+            true => {
+                if let Some(previous_named) = self.array_named_map.get(&self.peek_array_depth()) {
+                    match previous_named {
+                        true => Ok(()),
```

Comment:
These branches can be simplified:
* The `else` branches are unreachable. Because we already checked `self.array_named_map.entry(self.peek_array_depth())` before calling into this function.
* We can just check `current_named == previous_named`. When this fails, call `prev_token` unconditionally for `[`, and then once more `if current_named` for the keyword `ARRAY`.

Reply:
I change code to this one.
```
    fn check_same_named_array(&mut self, current_named: bool) -> Result<(), ParserError> {
        let previous_named = self.array_named_map.get(&self.peek_array_depth()).unwrap();
        if current_named != *previous_named {
            self.prev_token();
            if current_named {
                self.prev_token();
            }
            parser_err!(format!("syntax error at or near '{}'", self.peek_token()))?
        } else {
            Ok(())
        }
    }
```

Reply:
What about adding comments after each `prev_token()`, first `for [` and second `for keyword array`?